src/meta/processors/admin/Balancer.cpp

```diff
@@ -675,5 +794,84 @@ void Balancer::simplifyLeaderBalnacePlan(GraphSpaceID spaceId, LeaderBalancePlan
+    std::unique_ptr<kvstore::KVIterator> iter;
+    const auto& prefix = MetaServiceUtils::partPrefix(spaceId);
+    auto code = kv_->prefix(kDefaultSpaceId, kDefaultPartId, prefix, &iter);
+    if (code != kvstore::ResultCode::SUCCEEDED) {
+        LOG(ERROR) << "List Parts Failed";
+        return false;
+    }
+
+    std::unordered_map<HostAddr, int32_t> hostLoading;
+    while (iter->valid()) {
+        auto hosts = MetaServiceUtils::parsePartVal(iter->val());
+        for (auto& host : hosts) {
+            hostLoading[host]++;
+        }
+        iter->next();
+    }
+
+    for (auto it = hostLoading.begin(); it != hostLoading.end(); it++) {
+        int32_t min = it->second / replicaFactor;
+        auto pair = Bounds(min, min + 1);
+        LOG(INFO) << "Host: " << it->first << " Bounds: " << min << " : " << min + 1;
+        hostBounds_[it->first] = std::move(pair);
+    }
+    return true;
+}
+
+bool Balancer::calculationBounds(GraphSpaceID spaceId, int32_t replicaFactor) {
+    std::unique_ptr<kvstore::KVIterator> iter;
+    const auto& prefix = MetaServiceUtils::partPrefix(spaceId);
+    auto code = kv_->prefix(kDefaultSpaceId, kDefaultPartId, prefix, &iter);
+    if (code != kvstore::ResultCode::SUCCEEDED) {
+        LOG(ERROR) << "List Parts Failed";
+        return false;
+    }
+
+    std::unordered_map<HostAddr, int32_t> hostLoading;
+    PartAllocation partAddresses;
+    while (iter->valid()) {
+        auto partId = MetaServiceUtils::parsePartKeyPartId(iter->key());
+        auto hosts = MetaServiceUtils::parsePartVal(iter->val());
+        for (auto host : hosts) {
+            hostLoading[host]++;
+            partAddresses[partId].emplace_back(host);
+        }
+        iter->next();
+    }
+
+    std::unordered_map<PartitionID, std::vector<int32_t>> partLoading;
+    for (auto partIter = partAddresses.begin(); partIter != partAddresses.end(); partIter++) {
+        auto part = partIter->first;
+        for (auto address : partIter->second) {
+            auto hostIter = hostLoading.find(address);
+            if (hostIter == hostLoading.end()) {
+                LOG(ERROR) << "Can't find host " << address;
+                return false;
+            }
+
+            VLOG(3) << "Part: " << part << " size: " << hostIter->second;
+            partLoading[part].emplace_back(hostIter->second);
+        }
+    }
+
+    for (auto partIter = partLoading.begin(); partIter != partLoading.end(); partIter++) {
+        auto part = partIter->first;
+        auto size = partIter->second.size();
+        auto sum = std::accumulate(partIter->second.begin(), partIter->second.end(), 0);
+        int32_t min = sum / size / replicaFactor;
+        int32_t max = min;
+        if (sum % (size * replicaFactor) != 0) {
+            max += 1;
+        }
+
+        auto pair = Bounds(min, max);
+        LOG(INFO) << "Part: " << part << " Bounds: [" << pair.first << ", " << pair.second << "]";
+        partBounds_[part] = std::move(pair);
```

Comment:
Do you think the bound is too aggressive? Maybe a bigger range will be more safe. For example, given a part, three hosts hold it, I think the a range that covers average of the given three hosts looks more reasonable. `min < avg < max`. 

I am not pretty sure about that we should get it by `sum / size / replicaFactor`. A extreme example:
A: part [1]
B: part [1]
C: part [1-100]
They both have part 1, but we can't make promise that for the hosts which have part 1 should be in the range `sum / size / replicaFactor`, in this example, that would be `102 / 3 /3`.

Reply:
I‘m not sure `[0.95 * avg, 1.05 * avg]` is a bigger range.  Actually they are the same ([min, max])